connector/src/main/java/org/geysermc/connector/skin/SkinProvider.java

```diff
@@ -444,6 +466,62 @@ private static Skin supplyEars(Skin existingSkin, String earsUrl, EarsProvider p
         return data;
     }
 
+    /**
+     * If a skull has a username but no textures, request them.
+     * @param skullOwner the CompoundTag of the skull with no textures
+     * @return a completable GameProfile with textures included
+     */
+    public static CompletableFuture<GameProfile> requestTexturesFromUsername(CompoundTag skullOwner) {
+        return CompletableFuture.supplyAsync(() -> {
+            Tag uuidTag = skullOwner.get("Id");
+            String uuidToString = "";
+            JsonNode node;
+            GameProfile gameProfile = new GameProfile(UUID.randomUUID(), "");
+            boolean retrieveUuidFromInternet = !(uuidTag instanceof IntArrayTag); // also covers null check
+
+            if (!retrieveUuidFromInternet) {
+                int[] uuidAsArray = ((IntArrayTag) uuidTag).getValue();
+                // thank u viaversion
+                UUID uuid = new UUID((long) uuidAsArray[0] << 32 | ((long) uuidAsArray[1] & 0xFFFFFFFFL),
+                        (long) uuidAsArray[2] << 32 | ((long) uuidAsArray[3] & 0xFFFFFFFFL));
+                retrieveUuidFromInternet = uuid.version() != 4;
+                uuidToString = uuid.toString().replace("-", "");
+            }
+
+            try {
+                if (retrieveUuidFromInternet) {
+                    // Offline skin, or no present UUID
+                    HttpURLConnection con = (HttpURLConnection) new URL("https://api.mojang.com/users/profiles/minecraft/" + skullOwner.get("Name").getValue()).openConnection();
+                    con.setRequestProperty("User-Agent", "Geyser-" + GeyserConnector.getInstance().getPlatformType().toString() + "/" + GeyserConnector.VERSION);
+                    node = GeyserConnector.JSON_MAPPER.readTree(con.getInputStream());
```

Comment:
Would recommend moving this to the Http Utils class (same for the request a few lines down)